Src/Support/Google.Apis.Auth/OAuth2/AccessTokenWithHeaders.cs

```diff
@@ -66,9 +67,44 @@ internal AccessTokenWithHeaders(string token, string quotaProject = null)
         /// </summary>
         public IReadOnlyDictionary<string, IReadOnlyList<string>> Headers { get; }
 
-        internal void AddHeaders(HttpRequestHeaders requestHeaders)
+        /// <summary>
+        /// Builder class for <see cref="AccessTokenWithHeaders"/>.
+        /// </summary>
+        public class Builder
         {
-            foreach (var header in Headers)
+            /// <summary>
+            /// The GCP project ID used for quota and billing purposes. May be null.
+            /// </summary>
+            public string QuotaProject { get; set; }
+
+            /// <summary>
+            /// Builds and instance of <see cref="AccessTokenWithHeaders"/> with the given
+            /// token and the value set on this builder.
+            /// </summary>
+            /// <param name="token">The token to build the <see cref="AccessTokenWithHeaders"/> for.</param>
+            /// <returns>An <see cref="AccessTokenWithHeaders"/>.</returns>
+            public AccessTokenWithHeaders Build(string token) =>
+                new AccessTokenWithHeaders(token, QuotaProject);
+        }
+    }
+
+    /// <summary>
+    /// Extensions for <see cref="HttpRequestHeaders"/> that with help with 
+    /// including headers from an <see cref="AccessTokenWithHeaders"/> in a request.
+    /// </summary>
+    public static class HttpRequestHeadersExtensions
```

Comment:
Not as sure about this change - any reason not to make AddHeaders public in AccessTokenWithHeaders?
My reasoning is that we might want to have a more efficient internal structure, which may become inefficient if someone actually *uses* the `Headers` property, but which stays efficient if you just call `AddHeaders`. (Does that make sense?)

Basically, keeping the action inside the class gives us more implementation flexibility.

I suppose the extension method could always call a new internal method if we wanted to do that.

I suspect that for convenience it would also be helpful to have an extension method for `HttpRequest` as well, as I suspect 99% of callers will be `request.Headers.AddFrom`, rather than having access to the headers without the request.

Reply:
> any reason not to make AddHeaders public in AccessTokenWithHeaders?

It just felt cleaner this way, to keep the AccessTokenWithHeaders separated from where the token and headers end up, we don't have anything for adding the token to the headers (it can end up as part of the query string as well), and we don't have anything to add the token or the headers to gRPC metadata. I'm a happy to revert though, this is a design nit in any case.

I do see your point about not being able to depend on iternal implementation details unless we make those internal, but I wonder if efficiency is really going to be a concern here, we'll be addind very few headers, and even if users are writing their own credentials with headers, reallistically, how many headers can they be adding?

I will add a method for adding the headers to HttpRequest, extension or otherwise as we end up agreeing

Reply:
I think it would be simplest to keep it as it was before, and just make it public. Aside from anything else, it's probably more discoverable that way. (I doubt that most users will need to use it either way, mind you.)

Reply:
Done, and added a similar method that takes the HttpRequestMessage.